pkg/operations/kubernetesupgrade/upgrader.go

```diff
@@ -216,51 +215,83 @@ func (ku *Upgrader) upgradeAgentPools() error {
-			if err != nil {
-				ku.logger.Infof("Error deleting agent VM: %s\n", *vm.Name)
-				return err
-			}
+		// Create an auxiliary node if hasn't been created yet. This node is used to take on the load from deleting nodes.
+		// In a normal mode of operation the actual number of VMs in the pool [len(upgradeVMs)] is equal to agentCount.
+		// However, if the upgrade failed in the middle, the actual number of VMs might be agentCount+1 to include auxiliary node.
+		// In the later case we don't create an extra node.
+		if agentCount > 0 && len(upgradeVMs) == agentCount {
+			agentIndex := getAvailableIndex(upgradeVMs)
+			ku.logger.Infof("Adding auxiliary agent node with index %d", agentIndex)
 
 			err = upgradeAgentNode.CreateNode(*agentPool.Name, agentIndex)
 			if err != nil {
-				ku.logger.Infof("Error creating upgraded agent VM: %s\n", *vm.Name)
+				ku.logger.Infof("Error creating auxiliary node\n")
 				return err
 			}
+			// It is possible to reconstruct VM name, but we don't have to.
+			// The VM name is need for nodes that require upgrade. However this node already has desired orchestrator version.
+			upgradeVMs[agentIndex] = &UpgradeVM{"", true}
+		}
+
+		// Upgrade nodes in agent pool
+		upgradedCount, toBeUpgraded := 0, len(*agentPool.AgentVMs)
+		for agentIndex, vm := range upgradeVMs {
+			if vm.Upgraded {
+				continue
+			}
+			ku.logger.Infof("Upgrading Agent VM: %s, pool name: %s\n", vm.Name, *agentPool.Name)
 
-			err = upgradeAgentNode.Validate(vm.Name)
+			err := upgradeAgentNode.DeleteNode(&vm.Name)
 			if err != nil {
-				ku.logger.Infof("Error validating upgraded agent VM: %s, err: %v\n", *vm.Name, err)
+				ku.logger.Infof("Error deleting agent VM: %s\n", vm.Name)
 				return err
 			}
 
-			upgradedAgentsIndex[agentIndex] = true
+			// do not create last node in favor of auxiliary node
+			if upgradedCount == toBeUpgraded-1 {
+				ku.logger.Infof("Skipping creation of VM %s (indx %d) in favor of auxiliary node", vm.Name, agentIndex)
+				delete(upgradeVMs, agentIndex)
+			} else {
+				err = upgradeAgentNode.CreateNode(*agentPool.Name, agentIndex)
+				if err != nil {
+					ku.logger.Infof("Error creating upgraded agent VM: %s\n", vm.Name)
+					return err
+				}
+
+				err = upgradeAgentNode.Validate(&vm.Name)
+				if err != nil {
+					ku.logger.Infof("Error validating upgraded agent VM: %s, err: %v\n", vm.Name, err)
+					return err
+				}
+				vm.Upgraded = true
+			}
+			upgradedCount++
 		}
 
-		agentsToCreate := agentCount - len(upgradedAgentsIndex)
+		agentsToCreate := agentCount - upgradedCount
 		ku.logger.Infof("Expected agent count in the pool: %d, Creating %d more agents\n", agentCount, agentsToCreate)
 
 		// NOTE: this is NOT completely idempotent because it assumes that
 		// the OS disk has been deleted
 		for i := 0; i < agentsToCreate; i++ {
 			agentIndexToCreate := 0
-			for upgradedAgentsIndex[agentIndexToCreate] == true {
+			for upgradeVMs[agentIndexToCreate].Upgraded == true {
```

Comment:
part of me wonders if we have a bug here. Where if we are short vms we might not have a vm at an index. which would cause an index/key not found error and we aren't doing the safe way to check the map.  But I don't think this is going to change with this PR. @amanohar agree?

Reply:
I don't think this should be merged its written now and corner cases could easily result in error. Plus see my question above. I am not sure if adding upgradeVMs adds value.